# CREDIT CARD CUSTOMER CHURN ANALYSIS AND RETENTION DASHBOARD

## Objectives

* Fetch data from Kaggle and save as raw data", and  "engineer features for modelling"

## Inputs

* Write down which data or information you need to run the notebook 

## Outputs

* Write here which files, code or artefacts you generate by the end of the notebook 

## Additional Comments

* If you have any additional comments that don't fit in the previous bullets, please state them here. 



---

# Change working directory

* We are assuming you will store the notebooks in a subfolder, therefore when running the notebook in the editor, you will need to change the working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [12]:
import os
current_dir = os.getcwd()
current_dir

'\\\\talktalk\\redirectedfolders\\F.Afolabi\\Documents'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [13]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [14]:
current_dir = os.getcwd()
current_dir

'\\\\talktalk\\redirectedfolders\\F.Afolabi'

# Section 1

Section 1 content

In [21]:
import pandas as pd
import numpy as np
print(pd.__version__)





2.1.1


In [22]:
# Extracting the data 
df = pd.read_csv("../DataSet/Raw/BankChurners.csv")
df.shape

(10127, 23)

In [23]:
df.head()

,CLIENTNUM,Attrition_Flag,Customer_Age,Gender,Dependent_count,Education_Level,Marital_Status,Income_Category,Card_Category,Months_on_book,...,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt,Total_Trans_Ct,Total_Ct_Chng_Q4_Q1,Avg_Utilization_Ratio,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2
0,768805383,Existing Customer,45,M,3,High School,Married,$60K - $80K,Blue,39,...,12691.0,777,11914.0,1.335,1144,42,1.625,0.061,0.000093,0.99991
1,818770008,Existing Customer,49,F,5,Graduate,Single,Less than $40K,Blue,44,...,8256.0,864,7392.0,1.541,1291,33,3.714,0.105,0.000057,0.99994
2,713982108,Existing Customer,51,M,3,Graduate,Married,$80K - $120K,Blue,36,...,3418.0,0,3418.0,2.594,1887,20,2.333,0.000,0.000021,0.99998
3,769911858,Existing Customer,40,F,4,High School,Unknown,Less than $40K,Blue,34,...,3313.0,2517,796.0,1.405,1171,20,2.333,0.760,0.000134,0.99987
4,709106358,Existing Customer,40,M,3,Uneducated,Married,$60K - $80K,Blue,21,...,4716.0,0,4716.0,2.175,816,28,2.500,0.000,0.000022,0.99998


---

# Data Inspection; Data was inspected to understand the data, check missing values, spot any outliers, the type, the categories of the column object etc.


In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10127 entries, 0 to 10126
Data columns (total 23 columns):
 #   Column                                                                                                                              Non-Null Count  Dtype  
---  ------                                                                                                                              --------------  -----  
 0   CLIENTNUM                                                                                                                           10127 non-null  int64  
 1   Attrition_Flag                                                                                                                      10127 non-null  object 
 2   Customer_Age                                                                                                                        10127 non-null  int64  
 3   Gender                                                                           

## The Descriptive Statistics

In [25]:
df.describe()

,CLIENTNUM,Customer_Age,Dependent_count,Months_on_book,Total_Relationship_Count,Months_Inactive_12_mon,Contacts_Count_12_mon,Credit_Limit,Total_Revolving_Bal,Avg_Open_To_Buy,Total_Amt_Chng_Q4_Q1,Total_Trans_Amt,Total_Trans_Ct,Total_Ct_Chng_Q4_Q1,Avg_Utilization_Ratio,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1,Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2
count,1.012700e+04,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000,10127.000000
mean,7.391776e+08,46.325960,2.346203,35.928409,3.812580,2.341167,2.455317,8631.953698,1162.814061,7469.139637,0.759941,4404.086304,64.858695,0.712222,0.274894,0.159997,0.840003
std,3.690378e+07,8.016814,1.298908,7.986416,1.554408,1.010622,1.106225,9088.776650,814.987335,9090.685324,0.219207,3397.129254,23.472570,0.238086,0.275691,0.365301,0.365301
min,7.080821e+08,26.000000,0.000000,13.000000,1.000000,0.000000,0.000000,1438.300000,0.000000,3.000000,0.000000,510.000000,10.000000,0.000000,0.000000,0.000008,0.000420
25%,7.130368e+08,41.000000,1.000000,31.000000,3.000000,2.000000,2.000000,2555.000000,359.000000,1324.500000,0.631000,2155.500000,45.000000,0.582000,0.023000,0.000099,0.999660
50%,7.179264e+08,46.000000,2.000000,36.000000,4.000000,2.000000,2.000000,4549.000000,1276.000000,3474.000000,0.736000,3899.000000,67.000000,0.702000,0.176000,0.000181,0.999820
75%,7.731435e+08,52.000000,3.000000,40.000000,5.000000,3.000000,3.000000,11067.500000,1784.000000,9859.000000,0.859000,4741.000000,81.000000,0.818000,0.503000,0.000337,0.999900
max,8.283431e+08,73.000000,5.000000,56.000000,6.000000,6.000000,6.000000,34516.000000,2517.000000,34516.000000,3.397000,18484.000000,139.000000,3.714000,0.999000,0.999580,0.999990


In [26]:
df.isnull().sum()

CLIENTNUM                                                                                                                             0
Attrition_Flag                                                                                                                        0
Customer_Age                                                                                                                          0
Gender                                                                                                                                0
Dependent_count                                                                                                                       0
Education_Level                                                                                                                       0
Marital_Status                                                                                                                        0
Income_Category                                 

In [27]:
df.nunique()

CLIENTNUM                                                                                                                             10127
Attrition_Flag                                                                                                                            2
Customer_Age                                                                                                                             45
Gender                                                                                                                                    2
Dependent_count                                                                                                                           6
Education_Level                                                                                                                           7
Marital_Status                                                                                                                            4
Income_Category     

Section 2 content

In [28]:
# Checking for duplicates
df.duplicated().sum()    

0

Cardinality analysis revealed that teh customer identifier (CLIENTUM) contained fully unique values across all records, duplicate record checks no repeated observations 

TRANSFORMATION : I will make a copy of the dataset,  I will drop the 2 leakage columns(Naive_Bayes_Classifier), create the churn column, Encode categories, Handle Outliers 

In [30]:
# Making a copy of the original data frame as the  working copy for cleaning and transformation to preserve the original data frame for reference.
df_clean = df.copy()
df_clean.shape

(10127, 23)

In [ ]:
# Dropping the columns that are not useful for analysis and modeling.
columns_to_drop = ['Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_1', 'Naive_Bayes_Classifier_Attrition_Flag_Card_Category_Contacts_Count_12_mon_Dependent_count_Education_Level_Months_Inactive_12_mon_2']
df_clean.drop(columns=columns_to_drop, inplace=True)



In [36]:
df_clean.shape

(10127, 21)

In [42]:
# Creating a new column to indicate whether a customer is a churner or not based on the 'Attrition_Flag' column.
df_clean["Churn_Flag"] = df_clean["Attrition_Flag"].map({
    "Attrited Customer": 1,
    "Existing Customer": 0
})  

In [44]:
df_clean["Churn_Flag"].value_counts()
df_clean["Churn_Flag"].mean()

0.1606596227905599

The attrition label was transformed into binary churn indicator using encoding where churned customer were assigned 1 and existing customer a value of 0. The Churn rate is 16.1% . Out of the  10,127 customers: 16.1% have churned while 83.9% are retained. But the churn rate is still a significant revenue risk. 

In [46]:
# Creating Tenure Bands 
df_clean["Tenure_Band"] = pd.cut(df_clean["Months_on_book"], bins=[0, 12, 24, 36, 48, 60], labels=["0-1 year", "1-2 years", "2-3 years", "3-4 years", "4+ years"], include_lowest=True)


In [48]:
df_clean["Tenure_Band"].value_counts()

Tenure_Band
2-3 years    5418
3-4 years    3207
1-2 years     847
4+ years      655
0-1 year        0
Name: count, dtype: int64

In [50]:
# previewing the mapping 
df_clean[["Months_on_book", "Tenure_Band"]].head(10)

,Months_on_book,Tenure_Band
0,39,3-4 years
1,44,3-4 years
2,36,2-3 years
3,34,2-3 years
4,21,1-2 years
5,36,2-3 years
6,46,3-4 years
7,27,2-3 years
8,36,2-3 years
9,36,2-3 years


In [52]:
# Creating Age Bands
df_clean["Age_Band"] = pd.cut(df_clean["Customer_Age"], bins=[18, 30, 40, 50, 60, 100], labels=["18-30", "31-40", "41-50", "51-60", "60+"], include_lowest=True)
df_clean[["Customer_Age", "Age_Band"]].head(10)

,Customer_Age,Age_Band
0,45,41-50
1,49,41-50
2,51,51-60
3,40,31-40
4,40,31-40
5,44,41-50
6,51,51-60
7,32,31-40
8,37,31-40
9,48,41-50


In [56]:
# Creating Utilisation Segments 
df_clean["Utilisation_Segment"] = pd.cut(df_clean["Avg_Utilization_Ratio"], bins=[0, 0.3, 0.7, 1.0], labels=["Low", "Medium", "High",], include_lowest=True)

In [58]:
df_clean[["Avg_Utilization_Ratio", "Utilisation_Segment"]].head(10)
df_clean["Utilisation_Segment"].value_counts()

Utilisation_Segment
Low       6253
Medium    2815
High      1059
Name: count, dtype: int64

In [61]:
# Creating Activity_Level Segments
df_clean["Activity_Level"] = pd.cut(df_clean["Total_Trans_Ct"], bins=[0, 40, 80, 200], labels=["Low", "Medium", "High"], include_lowest=True)
df_clean[["Total_Trans_Ct", "Activity_Level"]].head(10)
df_clean["Activity_Level"].value_counts()

Activity_Level
Medium    5639
High      2550
Low       1938
Name: count, dtype: int64

In [63]:
# Creating spend_Drop flag
df_clean["Spend_Drop_Flag"] = (df_clean["Total_Amt_Chng_Q4_Q1"] < 1).astype(int )
df_clean[["Total_Amt_Chng_Q4_Q1", "Spend_Drop_Flag"]].head(10)
df_clean["Spend_Drop_Flag"].value_counts()

Spend_Drop_Flag
1    9127
0    1000
Name: count, dtype: int64

In [66]:
# Creating Transaction _Drop flag
df_clean["Transaction_Drop_Flag"] = (df_clean["Total_Ct_Chng_Q4_Q1"] < 1).astype(int )
df_clean[["Total_Ct_Chng_Q4_Q1", "Transaction_Drop_Flag"]].head(10)
df_clean["Transaction_Drop_Flag"].value_counts()    

Transaction_Drop_Flag
1    9293
0     834
Name: count, dtype: int64

In [68]:
# saving the cleaned and transformed data frame to a new csv file for analysis and modeling.
df_clean.to_csv("../DataSet/Cleaned/BankChurners_Cleaned.csv", index=False)     

The transformed dataset was exported to the Cleaned folder


In [70]:
df_BankChurn = df_clean.copy()
df_BankChurn.shape

(10127, 28)

In [72]:
df_BankChurn.columns

Index(['CLIENTNUM', 'Attrition_Flag', 'Customer_Age', 'Gender',
       'Dependent_count', 'Education_Level', 'Marital_Status',
       'Income_Category', 'Card_Category', 'Months_on_book',
       'Total_Relationship_Count', 'Months_Inactive_12_mon',
       'Contacts_Count_12_mon', 'Credit_Limit', 'Total_Revolving_Bal',
       'Avg_Open_To_Buy', 'Total_Amt_Chng_Q4_Q1', 'Total_Trans_Amt',
       'Total_Trans_Ct', 'Total_Ct_Chng_Q4_Q1', 'Avg_Utilization_Ratio',
       'Churn_Flag', 'Tenure_Band', 'Age_Band', 'Utilisation_Segment',
       'Activity_Level', 'Spend_Drop_Flag', 'Transaction_Drop_Flag'],
      dtype='object')

---

NOTE

* You may add as many sections as you want, as long as it supports your project workflow.
* All notebook's cells should be run top-down (you can't create a dynamic wherein a given point you need to go back to a previous cell to execute some task, like go back to a previous cell and refresh a variable content)

---

# Push files to Repo

* In cases where you don't need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.